minisat/core/Solver.cc

```diff
@@ -547,15 +557,20 @@ void Solver::reduceDB()
 
 void Solver::removeSatisfied(vec<CRef>& cs)
 {
-    int i, j;
-    for (i = j = 0; i < cs.size(); i++){
-        Clause& c = ca[cs[i]];
-        if (satisfied(c))
-            removeClause(cs[i]);
-        else
-            cs[j++] = cs[i];
+    auto i = cs.begin();
+    auto j = cs.begin();
+    auto end = cs.end();
+    while (i != end) {
```

Comment:
Just thinking... Are we sure that `cs.end()` is not affected by `removeClause(*i)`?

Reply:
Yes. There is obviously nothing that guarantees this for all refactorings across the future, but

1) Reading the code shows that `removeClause` can deallocate the memory allocated for the clause, but it will not affect its associated `CRef` or the vector storing those.
2) The test suite runs cleanly under Valgrind
3) The test suite runs cleanly under MSVC's debugging stdlib